lib/watchpool/watchpool.go

```diff
@@ -0,0 +1,169 @@
+		p.watchSets = make(map[uint64]*wsEntry)
+	}
+
+	// Whatever happens, we need to come back and decrement the ref count and
+	// possibly clean up at the end.
+	defer p.watcherDone(key)
+
+	// See if the key already exists in the share map
+	p.Lock()
+	entry, ok := p.watchSets[key]
+	if !ok {
+		// Create a new entry
+		entry = &wsEntry{
+			ws:   ws,
+			done: make(chan struct{}),
+			// Buffer in leaderChan is important to not block leader if it timesout
+			// with no followers. We rely on it to not deadlock the initial call too.
+			leaderDone: make(chan struct{}, 1),
+			refs:       1,
+		}
+		p.watchSets[key] = entry
+
+		// Trigger one of the Watch goroutines for this set to become the initial
+		// leader. It will likely be us in the select below but it could be another
+		// concurrent call that races. Note that if the select below takes a
+		// non-leader case because our context is cancelled for example, then either
+		// another Watcher will become leader instead, or all watchers will return
+		// for another reason and the state will be cleaned up by watcherDone
+		// preserving invariant that there is always one leader Watch for a given
+		// entry in the map. This won't block because we buffer this chan.
+		entry.leaderDone <- struct{}{}
+	} else {
+		// Add ourselves to the ref count
+		entry.refs++
+	}
+	// Unlock now before processing any more
+	p.Unlock()
+
+	// We are a follower
+	select {
+	case <-ctx.Done():
+		// Our context is done, just return it's error.
+		return ctx.Err()
+
+	case <-entry.done:
+		// The leader notified of a change in the watchset.
+		return nil
+
+	case <-entry.leaderDone:
+		// The current leader timed out (or we just setup the state above) and we
+		// won the race to become next leader. We need to watch until our context is
+		// done
+		watchErr := entry.ws.WatchCtx(ctx)
+		if watchErr == context.Canceled || watchErr == context.DeadlineExceeded {
+			// We timed out, pass the leader buck on to someone else (if there are any
+			// followers, if not this is harmless since it's a buffered chan).
+			entry.leaderDone <- struct{}{}
+		} else {
+			// An actual change happened, in this case notify followers and we can all
+			// be done so no need to pass on leadership.
+			close(entry.done)
+		}
+		return watchErr
+	}
+}
+
+func (p *WatchPool) watcherDone(key uint64) {
+	p.Lock()
+	defer p.Unlock()
+
+	entry, ok := p.watchSets[key]
+	if !ok || entry.refs < 1 {
+		// Shouldn't happen!
+		return
+	}
+	// Un-count watcher and possibly cleanup state.
+	entry.refs--
+	if entry.refs == 0 {
+		// Just remove the state. Any new watches will start a whole new leader etc.
+		delete(p.watchSets, key)
+	}
+}
+
+func (p *WatchPool) wsKey(ws memdb.WatchSet) (uint64, error) {
+	// Hash the map. hashstructure doesn't know how to hash a chan although
+	// reflect can do it without technically using anything unsafe via uintptr.
+	// Since memory addresses are unique already and we need them unordered since
```

Comment:
Docs for `reflect.Pointer` also say:

 > If v's Kind is Func, the returned pointer is an underlying code pointer, but not necessarily enough to identify a single function uniquely. The only guarantee is that the result is zero if and only if v is a nil func Value.

After confirming that Chans are one type this works for.

The _implication_ is that for non-Funcs, the pointer _is_ enough to identify a single instance uniquely.

Reply:
My reading of go src confirms this is the address of the `hchan` struct in the current runtime implementation which shouldn't be changed by GC even if the underlying buffer that elements are stored in is reallocated. If it is moved GC would need to update all pointer to it so I think it's safe to assume pointer comparison does what we expect even if it's not guaranteed by the spec.

https://github.com/golang/go/blob/master/src/reflect/value.go#L1373-L1374
https://github.com/golang/go/blob/master/src/reflect/value.go#L92
https://github.com/golang/go/blob/master/src/runtime/chan.go#L98-L102

Reply:
I've recently found I was wrong about this - GC _can_ reallocate and change pointers and then go back and change all the pointers that point to that object. This is stated explicitly in the unsafe package where it points out that storing a `uintptr` and later converted back and dereferenced is not safe since GC might have moved the object without knowing that the `uintptr` value needed to change. That's equivalent to this where a chan reallocation/move during GC would cause the same chan to later hash to a different value.

That said, it's still safe here since it's only an optimisation - we aren't dereferencing the pointer again later. The worst that would happen would be the hash change and now two or more requests watching the same chan would hash to different slots in the pool and duplicate effort again.

Assuming moves during GC are somewhat rare this means the optimisation will still work most of the time.